File: Winter-2019/stacks_queues/queue_from_stacks/queue_from_stacks.py
'''

The class definition for a Queue implemented with two stacks.

Here, lists are utilized as the data structure to represent the Stack.

In accordance with the nature of a stack, the list methods append() 
and pop() will be used exclusively.

'''
class TwoStackQueue:

    # the object's constructor
    def __init__(self):
        self.front = []
        self.back = []

    # public-facing method to add item to queue
    def push(self, item):
        self.back.append(item)
        if len(self.front) < 1:
            self._rebalance()
        return

    # public-facing method to pop item from queue
    def pop(self):
        # check for empty stacks first
        if self.size() == 0:
            print('Nothing to pop')
            return None
        else:
            item = self.front.pop()
            if len(self.front) < 1:
                self._rebalance()
            return item

    # public-facing helper method to get queue's size
    def size(self):
        return len(self.front) + len(self.back)

    # internal method to ensure the queue remains balanced
    # NOTE: this is the core logic of how to manage the two stacks
    # in this particular implementation.
    def _rebalance(self):
        # print("BEFORE REBAL: back: " + str(self.back) + " ... front: " + str(self.front))
        # This implementation only calls _rebalance() when the "front" 
        # stack is empty.
        while len(self.back) > 0:
            self.front.append(self.back.pop())
        # print("AFTER REBAL: back: " + str(self.back) + " ... front: " + str(self.front))
        return
    

# a driver function to test the class
def main():
    queue = TwoStackQueue()
    import pdb; pdb.set_trace()
    # expect a none-type object returned when popping from an empty queue
    print('pop: ' + str(queue.pop()))
    print('push 5')
    queue.push(5)
    print('push 6')
    queue.push(6)
    print('push 7')
    queue.push(7)
    print('push 8')
    queue.push(8)
    print('push 9')
    queue.push(9)
    print('push "a"')
    queue.push('a')
    print('size: ' + str(queue.size()))
    print('pop: ' + str(queue.pop()))
    print('push "4"')
    queue.push('4')
    print('pop: ' + str(queue.pop()))
    print('pop: ' + str(queue.pop()))
    print('pop: ' + str(queue.pop()))
    print('pop: ' + str(queue.pop()))
    print('pop: ' + str(queue.pop()))
    print('pop: ' + str(queue.pop()))
    print('size: ' + str(queue.size()))
    print('pop: ' + str(queue.pop()))
    return

main()

File: Winter-2019/stacks_queues/queue_from_stacks/queue_from_stacks.cpp
#include <string>
#include <stack>
#include <cassert>
#include <iostream>

using namespace std;

template<class T>
class TwoStackQueue {

    // public methods
    public:

        // constructor
        TwoStackQueue() {
            front = new stack<T>();
            back = new stack<T>();
        }

        // destructor
        ~TwoStackQueue() {
            delete front;
            delete back;
        }

        // push item into the queue
        void Push(T item) {
            back->push(item);
            if (front->size() == 0) {
                this->rebalance();
            }
        }

        // pop item from the queue
        bool Pop() {
            if (front->size() != 0) {
                // useful to see the item that is being popped
                // return val also indicates whether operation was successful
                cout << "popping item: " << front->top() << endl;
                front->pop();
                if (front->size() == 0) {
                    this->rebalance();
                }
                return true;
            }
            return false;
        }

        // get the queue's size
        int size() {
            return front->size() + back->size();
        }
                
    private:
        // data members
        stack<T>* front;
        stack<T>* back;

        // function to move queue items when front->stack is empty
        void rebalance() {
            while (back->size() > 0) {
                front->push(back->top());
                back->pop();
            }
        }
};

int main() {
    TwoStackQueue<string> strQueue;
    assert(strQueue.size() == 0);
    strQueue.Push("hello");
    strQueue.Push("world");
    assert(strQueue.size() == 2);
    bool success = strQueue.Pop();
    assert(success);
    strQueue.Push("on");
    strQueue.Push("fire");
    strQueue.Push("abc");
    strQueue.Push("xyz");
    assert(strQueue.size() == 5);
    success = strQueue.Pop();
    assert(success);
    success = strQueue.Pop();
    assert(success);
    success = strQueue.Pop();
    assert(success);
    success = strQueue.Pop();
    assert(success);
    success = strQueue.Pop();
    assert(success);
    success = strQueue.Pop();
    assert(!success);
    return 0;
}
